stan/math/rev/core/precomputed_gradients.hpp

```diff
@@ -17,51 +19,119 @@ namespace math {
+         0)...});
+  }
 
  public:
   /**
    * Construct a precomputed vari with the specified value,
-   * operands, and gradients.
+   * operands, gradients and optionally container operands and containers of
+   * gradients.
    *
+   * @tparam COps tuple of any container operands (var_value
+   * containing Eigen types)
+   * @tparam CGrads tupleof any container gradients (Eigen types)
    * @param[in] val The value of the variable.
    * @param[in] size Size of operands and gradients
    * @param[in] varis Operand implementations.
    * @param[in] gradients Gradients with respect to operands.
+   * @param container_operands any container operands
+   * @param container_gradients any container gradients
    */
-  precomputed_gradients_vari(double val, size_t size, vari** varis,
-                             double* gradients)
-      : vari(val), size_(size), varis_(varis), gradients_(gradients) {}
+  template <typename COps = std::tuple<>, typename CGrads = std::tuple<>>
+  precomputed_gradients_vari_template(double val, size_t size, vari** varis,
+                                      double* gradients,
+                                      COps&& container_operands
+                                      = std::tuple<>(),
+                                      CGrads&& container_gradients
+                                      = std::tuple<>())
+      : vari(val),
+        size_(size),
+        varis_(varis),
+        gradients_(gradients),
+        container_operands_(std::forward<COps>(container_operands)),
+        container_gradients_(std::forward<CGrads>(container_gradients)) {
+    check_sizes(std::make_index_sequence<N_containers>());
+  }
 
   /**
    * Construct a precomputed vari with the specified value,
-   * operands, and gradients.
+   * operands, gradients and optionally container operands and containers of
+   * gradients.
    *
    * @tparam Arith An arithmetic type
    * @tparam VecVar A vector of vars
    * @tparam VecArith A vector of arithmetic types
+   * @tparam COps tuple of any container operands (var_value
+   * containing Eigen types)
+   * @tparam CGrads tupleof any container gradients (Eigen types)
    * @param[in] val The value of the variable.
    * @param[in] vars Vector of operands.
    * @param[in] gradients Vector of partial derivatives of value
    * with respect to operands.
+   * @param container_operands any container operands
+   * @param container_gradients any container gradients
    * @throws std::invalid_argument if the sizes of the vectors
    * don't match.
    */
-  template <typename Arith, typename VecVar, typename VecArith>
-  precomputed_gradients_vari(Arith val, VecVar&& vars, VecArith&& gradients)
+  template <typename Arith, typename VecVar, typename VecArith,
+            typename COps = std::tuple<>, typename CGrads = std::tuple<>,
+            require_all_vector_t<VecVar, VecArith>* = nullptr>
+  precomputed_gradients_vari_template(Arith val, const VecVar& vars,
```

Comment:
Should we just call this class `precomputed_gradients_impl`?

Reply:
`_impl`s are usually things, hidden in internal namespace that noone, except implementor needs to know about. This class could be used directly. And it actually differs from `precomputed_gradients_vari` by the fact of being a template.